lib/spack/spack/spec.py

```diff
@@ -1248,18 +1449,25 @@ def _add_dependency(self, spec, deptypes):
             raise DuplicateDependencyError(
                 "Cannot depend on '%s' twice" % spec)
 
-        # create an edge and add to parent and child
-        dspec = DependencySpec(self, spec, deptypes)
-        self._dependencies[spec.name] = dspec
-        spec._dependents[self.name] = dspec
+        self.add_dependency_edge(spec, deptypes)
+
+    def add_dependency_edge(self, dependency_spec, deptype):
```

Comment:
Can this just be called `add_dependency`?  Rationale:

1. To some extent, the edges are an abstract concept to the user of `Spec`, though we do have methods that can iterate over edges.
2. But, this function doesn't take an edge.  It take a `Spec` -- i.e., the actual dependency.  So I think `edge` here can be confusing to the caller.

Reply:
I'm just now noticing that `_add_dependency` still exists and just calls this.  I think these should be consolidated.

Reply:
> I'm just now noticing that _add_dependency still exists and just calls this. I think these should be consolidated to add_dependency

`_add_dependency` is there just to enforce the constraint that we can't have two dependencies with the same name at the moment. It will be removed as soon as we arrive at separate concretization of build deps. 

For the naming of `add_dependency_edge`  I have no preference. My intention with having `edge` in the name was to give a hint that:
1. The dependency spec (the DAG node node) will not be changed in any way by this call
2. All the properties that are added (deptypes and in the future virtuals etc.) are associated with an edge

Reply:
@tgamblin This renaming is still missing. If the Pr seems good enough I can follow with a smaller PR that does just the renaming. Otherwise, what should be the new names? I was thinking:
- `_add_dependency` -> `legacy_add_dependency` (this will be removed when we add separate concretization of build deps)
- `add_dependency_edge` -> `add_dependency`